### PR TITLE
Refactor and document DisplaySourceSet

### PR DESCRIPTION
**Pull request by qwwdfsad**
* Decommission `SelfRepresentingSingletonSet` as harmful
* Document the purpose and the semantics of `DisplaySourceSet`
* Slightly refactor code around

### CODE REVIEW

File: core/src/main/kotlin/model/DisplaySourceSet.kt [hunk 1]
```diff
@@ -1,32 +1,56 @@
...
+ * @property sourceSetIDs unique stable id of the display source set.
+ *  It is composite by definition, as it uniquely defines the source set and all nested source sets.
+ *  Apart from names, it also contains a substitute to a full source set path in order to differentiate
+ *  source sets with the same name in a stable manner.
+ * @property name corresponds to the name of the original [DokkaSourceSet]
+ * @property platform the platform of the source set. If the source set is a mix of multiple source sets
+ *  that correspond to multiple KMP platforms, then it is [Platform.common]
  */
-data class DisplaySourceSet(
+public data class DisplaySourceSet(
```
Comment (qwwdfsad) [inline core/src/main/kotlin/model/DisplaySourceSet.kt:26]: IMO it is worth making all constructors `internal` and leaving only `DokkaSourceSet.toDisplaySourceSet`. The API already kind of dictate that, but I'm not sure how drastic this change will be

**Comment by qwwdfsad** [comment]
The integration test failure is related to pathsaver plugin in Knit -- I'll see if it's patchable locally

**Comment by qwwdfsad** [comment]
I've patched the Knit, updated it in coroutines/serialization and checked it helps. Next step is to update the submodule

**Review by vmishenev — APPROVED** [review]

File: core/src/main/kotlin/model/DisplaySourceSet.kt (hunk 1 above)
Comment (qwwdfsad) [reply core/src/main/kotlin/model/DisplaySourceSet.kt:27]: Probably worth figuring it out systematically. This one might be too strict, especially for 1.9.20

**Pull request merged**